ddht/v5_1/alexandria/content_storage.py

```diff
@@ -55,6 +162,18 @@ def enumerate_keys(
 
         yield from all_keys[left:right]
 
+    @contextlib.contextmanager
+    def atomic(self) -> Iterator[ContentStorageAPI]:
+        batch = _AtomicBatch(self)
+
+        yield batch
+
+        to_delete, to_write = batch.finalize()
+        for content_key in to_delete:
+            self.delete_content(content_key)
+        for content_key, content in to_write:
+            self.set_content(content_key, content, exists_ok=True)
```

Comment:
It seems like maybe this is as "atomic" as we get? What are the chances/consequences if something goes wrong while iterating through these?

Reply:
Yeah, I'll add a comment.  I considered options but didn't see any ways that I could actually make this truly atomic.